tasks/storage-market/storage_market.go

```diff
@@ -0,0 +1,614 @@
+			if goUrl.Scheme == "pieceref" {
+				refNum, err := strconv.ParseInt(goUrl.Opaque, 10, 64)
+				if err != nil {
+					return xerrors.Errorf("UUID: %s parsing piece reference number: %w", deal.UUID, err)
+				}
+
+				var complete bool
+				err = d.db.QueryRow(ctx, `SELECT pp.complete
+												FROM parked_pieces pp
+												JOIN parked_piece_refs ppr ON pp.id = ppr.piece_id
+												WHERE ppr.ref_id = $1;`, refNum).Scan(&complete)
+				if err != nil {
+					return xerrors.Errorf("UUID: %s getting piece park status: %w", deal.UUID, err)
+				}
+
+				if complete {
+					deal.Started = true
+					_, err = d.db.Exec(ctx, `UPDATE market_mk12_deal_pipeline SET started = TRUE WHERE uuid = $1`, deal.UUID)
+					if err != nil {
+						return xerrors.Errorf("failed to mark deal %s as started: %w", deal.UUID, err)
+					}
+					log.Infof("UUID: %s deal started successfully", deal.UUID)
+					return nil
+				}
+			}
+		} else {
+			// If no URL found for offline deal then we should try to find one
+			if deal.Offline {
+				found, err := d.findURLForOfflineDeals(ctx, deal.UUID, deal.PieceCid)
+				if err != nil {
+					return err
+				}
+				if found {
+					_, err = d.db.Exec(ctx, `UPDATE market_mk12_deal_pipeline SET started = TRUE WHERE uuid = $1`, deal.UUID)
+					if err != nil {
+						return xerrors.Errorf("UUID: %s: failed to mark deal as started: %w", deal.UUID, err)
+					}
+					log.Infof("UUID: %s deal started successfully", deal.UUID)
+					return nil
+				}
+			}
+		}
+	}
+
+	// Create commP task
+	if deal.Started && !deal.AfterCommp && deal.CommTaskID == nil {
+		// Skip commP is configured to do so
+		if d.cfg.Market.StorageMarketConfig.MK12.SkipCommP {
+			_, err := d.db.Exec(ctx, `UPDATE market_mk12_deal_pipeline SET after_commp = TRUE, commp_task_id = NULL WHERE uuid = $1`, deal.UUID)
+			if err != nil {
+				return xerrors.Errorf("UUID: %s: updating deal pipeline: %w", deal.UUID, err)
+			}
+			log.Infof("UUID: %s: commP skipped successfully", deal.UUID)
+			return nil
+		}
+
+		d.pollers[pollerCommP].Val(ctx)(func(id harmonytask.TaskID, tx *harmonydb.Tx) (shouldCommit bool, err error) {
+			// update
+			n, err := tx.Exec(`UPDATE market_mk12_deal_pipeline SET commp_task_id = $1 WHERE uuid = $2 AND commp_task_id IS NULL`, id, deal.UUID)
+			if err != nil {
+				return false, xerrors.Errorf("UUID: %s: updating deal pipeline: %w", deal.UUID, err)
+			}
+
+			// commit only if we updated the piece
+			return n > 0, nil
+		})
+		log.Infof("UUID: %s: commP task created successfully", deal.UUID)
+		return nil
+	}
+
+	// Create Find Deal task
+	if deal.Started && deal.AfterCommp && deal.AfterPSD && !deal.AfterFindDeal && deal.FindDealTaskID == nil {
+		d.pollers[pollerFindDeal].Val(ctx)(func(id harmonytask.TaskID, tx *harmonydb.Tx) (shouldCommit bool, err error) {
+			// update
+			n, err := tx.Exec(`UPDATE market_mk12_deal_pipeline SET find_deal_task_id = $1 WHERE uuid = $2 AND find_deal_task_id IS NULL`, id, deal.UUID)
```

Comment:
But we want this to be after_psd. PSD sends the message and this task attached the on chain deal IDs to deals. I also have if check above to ensure task is created for deals in certain condition only.

Reply:
Ah, I meant that the data in the `deal.AfterPSD && !deal.AfterFindDeal` check may have changed in the DB by the time we get to this UPDATE query, so this update should have a WHERE condition equivalent to the if above (basically if two pollers run at the same time, and get to this point we'd end up with two tasks created, one failing because it was overwritten in the pipeline)

Reply:
I have made these checks more atomic by wrapping them in transactions.